### ISSUE TITLE
Result statistics broken in new version of dbt3-post-process

### ISSUE THREAD

**Issue by olepbr** [osdldbt/dbt3#1]
Hi,

Thanks for maintaining dbt3!

I've been using it to benchmark PostgreSQL over the past few weeks and I just noticed that the generated statistics in `summary.rst` don't make sense. As an example:

```rst
========  ========  ========  ========  ========  ========  ========  ========
 Stream      Q1        Q2        Q3        Q4        Q5        Q6        Q7
========  ========  ========  ========  ========  ========  ========  ========
       1      1.73      0.26      0.41      0.13      0.79      0.35      1.00
       2      1.74      0.25      0.40      0.14      0.79      0.36      0.97
     Min      1.74      0.25      0.40      0.14      0.79      0.36      0.97
     Max      1.74      0.25      0.40      0.14      0.79      0.36      0.97
     Avg      0.87      0.12      0.20      0.07      0.39      0.18      0.48
========  ========  ========  ========  ========  ========  ========  ========

```

As you can see, both Min and Max are simply the value of the last stream, whilst Avg is the last stream divided by the number of streams. As an example, Q1 here is reported to have an average of 0.87 instead of the correct 1.74. (The same type of errors can be seen in the example in `doc/dbt3.rst`.)

If we take a look at what the script does, the error is evident:

https://github.com/osdldbt/dbt3/blob/a668b1dc43efc4cd2e80adf2a14bc0c5f23536be/scripts/dbt3-post-process.in#L199-L218

`STREAM` is set in a for loop further up in the script, but this section is not contained within it. Thus, `STREAM` just retains its latest value (in this case 2) and the basis of the calculation becomes just the final stream instead of all streams.

Looking at the history for `scripts/dbt3-post-process.in`, I believe this error was introduced when the post-processing was converted from R to shell in a1719d7.

I think a solution would be to not use the `STREAM` variable in the `STRING` regex for the aggregations, e.g.

```sh
STRING="^PERF\.THRUPUT\.QS[[:digit:]]\+\.Q${Q},"
```

Happy to make a PR if you agree!

All the best,
Ole

**Comment by markwkm** [comment]
Hi!

I haven't spent enough time to figure out where the bug is, but to restate what I think the facts are, `STREAMS` (the plural) should be `2` and `STREAM` (singular) should be the current stream (`1` or `2`).  My initial reaction is that line 217 isn't wrong, but perhaps there is a typo elsewhere so that `STREAMS` is getting clobbered?

I'll hopefully have more time to look next week, but I hope that might be helpful if that might illuminate where the source of the issue is...

**Comment by olepbr** [comment]
> I haven't spent enough time to figure out where the bug is, but to restate what I think the facts are, `STREAMS` (the plural) should be `2` and `STREAM` (singular) should be the current stream (`1` or `2`). My initial reaction is that line 217 isn't wrong, but perhaps there is a typo elsewhere so that `STREAMS` is getting clobbered?

Sorry, I was perhaps a bit unclear in my initial report. `STREAM`and `STREAMS` are both fine (and I agree with your definitions of them), but the error isn't on line 217, it's on 207 (and 184 and 162 for the other aggregates). The problem is that we can't set `STRING` using `STREAM` (singular) when computing the aggregates, because then the shell pipe only gets one line to operate on (that of the last stream) instead of a set of lines (one per stream).

> I'll hopefully have more time to look next week, but I hope that might be helpful if that might illuminate where the source of the issue is...

No worries! FWIW, I tested changing the three mentioned lines to my proposed regex and the resulting statistics seem correct to me :-)

**Comment by markwkm** [comment]
Aha, please go ahead and open a PR.

**Comment by markwkm** [comment]
I finally made more time to look and I understand now.  So min and max were broken too when calculating the query stats.  I took a slightly different approach since there could be more than 9 streams.  And I tightened up the regex for the refresh stream calculations too.  I'll make time soon to push a new release.

```diff
diff --git a/scripts/dbt3-post-process.in b/scripts/dbt3-post-process.in
index cf27b0b..3cb48ef 100755
--- a/scripts/dbt3-post-process.in
+++ b/scripts/dbt3-post-process.in
@@ -155,11 +155,11 @@ throughput_query() {
    printf "     Min"
    for Q in $(seq "${Q0}" "${Q1}"); do
        if [ "${Q}" -eq 23 ]; then
-           STRING="^PERF\.THRUPUT\.RFST.*\.RF1,"
+         STRING="^PERF\.THRUPUT\.RFST[0-9]\+\.RF1,"
        elif [ "${Q}" -eq 24 ]; then
-           STRING="^PERF\.THRUPUT\.RFST.*\.RF2,"
+         STRING="^PERF\.THRUPUT\.RFST[0-9]\+\.RF2,"
        else
-           STRING="^PERF\.THRUPUT\.QS${STREAM}\.Q${Q},"
+         STRING="^PERF\.THRUPUT\.QS[0-9]\+\.Q${Q},"
        fi

        if [ $HTML -eq 1 ]; then
@@ -177,11 +177,11 @@ throughput_query() {
    printf "     Max"
    for Q in $(seq "${Q0}" "${Q1}"); do
        if [ "${Q}" -eq 23 ]; then
-           STRING="^PERF\.THRUPUT\.RFST.*\.RF1,"
+         STRING="^PERF\.THRUPUT\.RFST[0-9]\+\.RF1,"
        elif [ "${Q}" -eq 24 ]; then
-           STRING="^PERF\.THRUPUT\.RFST.*\.RF2,"
+         STRING="^PERF\.THRUPUT\.RFST[0-9]\+\.RF2,"
        else
-           STRING="^PERF\.THRUPUT\.QS${STREAM}\.Q${Q},"
+         STRING="^PERF\.THRUPUT\.QS[0-9]\+\.Q${Q},"
        fi

        if [ $HTML -eq 1 ]; then
@@ -199,11 +199,11 @@ throughput_query() {
    printf "     Avg"
    for Q in $(seq "${Q0}" "${Q1}"); do
        if [ "${Q}" -eq 23 ]; then
-           STRING="^PERF\.THRUPUT\.RFST.*\.RF1,"
+         STRING="^PERF\.THRUPUT\.RFST[0-9]\+\.RF1,"
        elif [ "${Q}" -eq 24 ]; then
-           STRING="^PERF\.THRUPUT\.RFST.*\.RF2,"
+         STRING="^PERF\.THRUPUT\.RFST[0-9]\+\.RF2,"
        else
-           STRING="^PERF\.THRUPUT\.QS${STREAM}\.Q${Q},"
+         STRING="^PERF\.THRUPUT\.QS[0-9]\+\.Q${Q},"
        fi

        if [ $HTML -eq 1 ]; then
```

**Comment by markwkm** [comment]
Fixed in 66bead40647f5e7585c260c30138954c4e39fe80 and it's in v2.0.5.

**Comment by olepbr** [comment]
Thank you so much for fixing this! Sorry for the poor response time on my end, scrambling to finish my master's…
